## Trimming and assembly

The data for Tympa and Octomys each were generated by multiplexing six RNA libraries (liver, heart, muscle, lung, kidney, and gonad (testis or ovary) on one lane.  This creates concerns about lane effects, which we will just have to include as a source of noise.

First we used Trimmomatic to trim the reads; both sets of transcriptomes were trimmed with trimmomatic version 0.32. For Tympa I ran each command line independently:

```
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Heart_log.txt Heart_R1.fastq.gz Heart_R2.fastq.gz Heart_R1_trim_paired.fastq.gz Heart_R1_trim_single.fastq.gz Heart_R2_trim_paired.fastq.gz Heart_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Kidney_log.txt Kidney_R1.fastq.gz Kidney_R2.fastq.gz Kidney_R1_trim_paired.fastq.gz Kidney_R1_trim_single.fastq.gz Kidney_R2_trim_paired.fastq.gz Kidney_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Liver_log.txt Liver_R1.fastq.gz Liver_R2.fastq.gz Liver_R1_trim_paired.fastq.gz Liver_R1_trim_single.fastq.gz Liver_R2_trim_paired.fastq.gz Liver_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Lung_log.txt Lung_R1.fastq.gz Lung_R2.fastq.gz Lung_R1_trim_paired.fastq.gz Lung_R1_trim_single.fastq.gz Lung_R2_trim_paired.fastq.gz Lung_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Muscle_log.txt Muscle_R1.fastq.gz Muscle_R2.fastq.gz Muscle_R1_trim_paired.fastq.gz Muscle_R1_trim_single.fastq.gz Muscle_R2_trim_paired.fastq.gz Muscle_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Testis_log.txt Testis_R1.fastq.gz Testis_R2.fastq.gz Testis_R1_trim_paired.fastq.gz Testis_R1_trim_single.fastq.gz Testis_R2_trim_paired.fastq.gz Testis_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
```
 For Octomys, I wrote a bash script:

``` bash
#!/bin/bash

files=”Heart
Liver
Lung
Muscle
Kidney
Ovary”

for file in $files
do
	java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog ${file}_log.txt AO248_${file}_R1.fastq.gz AO248_${file}_R2.fastq.gz AO248_${file}_R1_trim_paired.fastq.gz AO248_${file}_R1_trim_single.fastq.gz AO248_${file}_R2_trim_paired.fastq.gz AO248_${file}_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
done
```
And for Xenopus (BJE3909_tropicalis and BJE4168_laevis), here are the commands in iqaluk:

```
cat *R1* > BJE3909cDNA_R1.fastq.gz
cat *R2* > BJE3909cDNA_R2.fastq.gz
cat *R1* > BJE4168cDNA_R1.fastq.gz
cat *R2* > BJE4168cDNA_R2.fastq.gz

java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog BJE3909cDNA_R1.fastq.gz_log.txt BJE3909cDNA_R1.fastq.gz BJE3909cDNA_R2.fastq.gz BJE3909cDNA_R1_trim_paired.fastq.gz BJE3909cDNA_R1_trim_single.fastq.gz BJE3909cDNA_R2_trim_paired.fastq.gz BJE3909cDNA_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36

java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog BJE4168cDNA_R1.fastq.gz_log.txt BJE4168cDNA_R1.fastq.gz BJE4168cDNA_R2.fastq.gz BJE4168cDNA_R1_trim_paired.fastq.gz BJE4168cDNA_R1_trim_single.fastq.gz BJE4168cDNA_R2_trim_paired.fastq.gz BJE4168cDNA_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36

```


I am going to use Trinity version 2.1.1 for assembly.  I am using only reads where both pairs passed the trimming step and I am going to concatenate forward and reverse reads from all libraries for each species and do the assembly jointly for all libraries for each species. This should reduce redundancy, facilitate read mapping, and hopefully simplify things in general.

On info, here is the concatenation command for the forward (R1) reads for Octomys:

```
cat Heart/AO248_Heart_R1_trim_paired.fastq Kidney/AO248_Kidney_R1_trim_paired.fastq Liver/AO248_Liver_R1_trim_paired.fastq Lung/AO248_Lung_R1_trim_paired.fastq Muscle/AO248_Muscle_R1_trim_paired.fastq Ovary/AO248_Ovary_R1_trim_paired.fastq > AO248_all_R1_trim_paired.fastq
```

and for the reverse (R2) reads:

```
cat Heart/AO248_Heart_R2_trim_paired.fastq Kidney/AO248_Kidney_R2_trim_paired.fastq Liver/AO248_Liver_R2_trim_paired.fastq Lung/AO248_Lung_R2_trim_paired.fastq Muscle/AO248_Muscle_R2_trim_paired.fastq Ovary/AO248_Ovary_R2_trim_paired.fastq > AO248_all_R2_trim_paired.fastq
```

And now the trinity assembly using version 2.1.1:

```
/home/ben/trinityrnaseq-2.1.1/Trinity --seqType fq --left AO248_all_R1_trim_paired.fastq.gz --right AO248_all_R2_trim_paired.fastq.gz --CPU 6 --max_memory 20G
```
and for the concatenated tympa transcriptomes:

```
/home/ben/trinityrnaseq-2.1.1/Trinity --seqType fq --left tympa_all_R1_trim_paired.fastq.gz --right tympa_all_R2_trim_paired.fastq.gz --CPU 6 --max_memory 20G
```

Update in July, 2016.  I am working with the Octomys and Tympanoctomys trinity assemblies again in order to provide Nate with some preliminary data for his proposal.  I have renamed the trinity.fasta files as follows: 
'/home/ben/2014_Tympanoctomys_transcriptomes/Tympano/Tympano_joint_trinity_assembly_with_concatenated_reads/trinity_out_dir/Tympa_all_transcriptomes_assembled_together.fasta'
'/home/ben/2014_Tympanoctomys_transcriptomes/Octomys/Octomys_joint_trinity_assembly_with_concatenated_reads/trinity_out_dir/Octomys_all_transcriptomes_assembled_together.fasta'

These have 280,712 and 308,854 assembled sequences for T. barrarae and O. mimax, respectively.  Then I used cd-hit-est to reduce redundancy as follows:
```
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Tympa_all_transcriptomes_assembled_together.fasta -o ./cdhit_est/Tympa_all_transcriptomes_assembled_together_unique.fasta -c 1.00 -n 10 -M 2000 -T 8
```
and 

```
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Octomys_all_transcriptomes_assembled_together.fasta -o ./cdhit_est/Octomys_all_transcriptomes_assembled_together_unique.fasta -c 1.00 -n 10 -M 2000 -T 8
```

And for XL and XT

```
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Assembled_BJE4168_Laevis_Trinity.fasta -o Assembled_BJE4168_Laevis_Trinity_unique.fasta -c 1.00 -n 10 -M 2000 -T 8

/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Assembled_BJE3909_Tropicalis_Trinity.fasta -o Assembled_BJE3909_Tropicalis_Trinity_unique.fasta -c 1.00 -n 10 -M 2000 -T 8
```


Here are some old commands for the ST and XL unigene dbs:

```
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Str.seq.uniq -o Str.seq.uniq_cdhitest100.fasta -c 1.00 -n 10 -M 2000 -T 8
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Str.seq.uniq_cdhitest100.fasta -o Str.seq.uniq_cdhitest80.fasta -c 0.80 -n 5 -M 2000 -T 8
```
and
```
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Xl.seq.uniq -o Xl.seq.uniq_cdhitest100.fasta -c 1.00 -n 10 -M 2000 -T 8
/usr/local/cdhit/cd-hit-v4.6.1-2012-08-27/cd-hit-est -i Xl.seq.uniq_cdhitest100.fasta -o Xl.seq.uniq_cdhitest80.fasta -c 0.80 -n 5 -M 2000 -T 8
```

I used RepARC.pl to use a kmer approach to identify and assemble transposible elements from the raw (trimmed) sequences. This did not work initially so I tried it again with a threshold of 70 as follows:
```
./RepARK.pl -l AO248_all_R1_trim_paired.fastq.gz -l AO248_all_R2_trim_paired.fastq.gz -t 70
```
```
./RepARK.pl -l tympa_all_R1_trim_paired.fastq.gz -l tympa_all_R2_trim_paired.fastq.gz -t 70
```
This worked and made a file in a folder called 'velvet_repeat_lib' which was called 'contigs.fa' which had the repeat sequences that were constructed from the kmers.  Then this can be fed into TEclass.  I had to log into evanslab and copy the contig.fa files there because I got a perl error associated with a missing module on my account.  I was unable to use cpan to install it because of priviledges. Anyhow, here is how it was run on evanslab from within the TEclass-2.1.3 directory:
```
/home/evanslab/Hymeno_fastqc/RepArk_analysis/TEclass-2.1.3/TEclassTest.pl -r ./Octomys/contigs.fa -o ./Octomys/TE_count_and_categories.out
```
and
```
/home/evanslab/Hymeno_fastqc/RepArk_analysis/TEclass-2.1.3/TEclassTest.pl -r ./Tympa/contigs.fa -o ./Tympa/TE_count_and_categories.out
```
Results
```
tympa
Repeat statistics:
DNA transposons: 13041
LTRs:  19229
LINEs: 13333
SINEs: 5874
Unclear:  16624

Octomys
Repeat statistics:
DNA transposons: 11885
LTRs:  18229
LINEs: 13734
SINEs: 6151
Unclear:  16333
```

Some directories were made for each analysis with weird names but neither output matched the file of the input line. I figured out which was which by comparing the '.stat' file to the output from each analysis.

Update: these results are bogus.  The manual suggests that this should not be done with whole genome sequences becaise everything will be assinged to a category.  Better to just feed the unique seqs into RepeatMasker, which can be done like this:

```
/usr/local/RepeatMasker/RepeatMasker -q -species rodentia -noisy Octomys_all_transcriptomes_assembled_together_unique.fastar
/usr/local/RepeatMasker/RepeatMasker -q -species rodentia -noisy Tympa_all_transcriptomes_assembled_together_unique.fasta
```

the `-q` flag does a quick search.  If this takes too long, do `-qq` instead.


I had an idea to use cd-hit to progressively simplify the transcriptome assembly to see if we can get a precipitous drop in read number because paralogs are combined below a certain similarity threshold.  To demonstrate this effect, I downloaded XL and ST unigene dbs and am trying cdhit on them.  First I'll do it at 100% and then I'll feed this into cdhit again at 99, 95, 90 and 80.

```
/home/ben/cd-hit-v4.6.1-2012-08-27/cd-hit -i Xl.seq.uniq -o Xl.seq.uniq_cdhit100.fasta -c 1.00 -n 5 -M 2000
/home/ben/cd-hit-v4.6.1-2012-08-27/cd-hit -i Str.seq.uniq -o Str.seq.uniq_cdhit100.fasta -c 1.00 -n 5 -M 2000
```
This ended up not working, so I have abandoned the cd-hit-est approach

The kmer approach using RepArk.pl used a default kmer size of 31 bases.  One possibility is that this size is too small to effectively distinguish a 'peak' of abundance of kmers.  So I am going to try a larger kmer size as follows:

```
./RepARK.pl -l AO248_all_R1_trim_paired.fastq.gz -l AO248_all_R2_trim_paired.fastq.gz -k 70 -o repArc_kmer_70
```
```
./RepARK.pl -l tympa_all_R1_trim_paired.fastq.gz -l tympa_all_R2_trim_paired.fastq.gz -k 70 -o repArc_kmer_70
```
The kmer densities can be plotted using this R script:

```R
library(ggplot2)

# make a pdf
pdf("kmer_density plot.pdf",w=6, h=2, version="1.4", bg="transparent")

# load the data
tymp<-read.table("jf_RepARK.tymp_histo")
tymp$species <- 'tymp'
oct<-read.table("jf_RepARK.octomys_histo")
oct$species <- 'oct'

# combine the data
data <- rbind(tymp, oct)

# make a density plot (data$V2 has the counts of each kmer)
ggplot(data, aes(V2, fill = species)) +
  # make it transparent and add a limit to the X and Y axes for clarity
  geom_density(alpha = 0.2) + xlim(0,5000) + ylim(0,0.002) +
  # modify the labels
  xlab("Occurance") + ylab("Density") +
  # modify the title
  ggtitle(expression(paste("Densities of 31-mers for ",italic("T. barrerae")," and ",italic("O. mimax")," transcriptomes"))) +
  # get rid of the background
  theme_classic() +
  # fix the legend
  scale_fill_manual(values=c("red", "blue"),
                    name="Species",
                    labels=c(expression(paste(italic("O. mimax"))), expression(paste(italic("T. barrerae")))))+
  # get the legend to be left justified
  theme(legend.text.align	=0) +
  # move the legend over
  theme(legend.position = c(.8, .5))+
  # make the title smaller
  theme(plot.title = element_text(size = 10)) +
  # make legend title smaller too
  theme(legend.title = element_text(size = 8)) +
  # make legend text smaller too
  theme(legend.text = element_text(size = 8)) +
  # make x axis smaller too
  theme(axis.title.x = element_text(size = 8)) +
  # make y axis smaller too
  theme(axis.title.y = element_text(size = 8)) +
  # make axis text smaller too
  theme(axis.text = element_text(size = 8)) +
  geom_segment(aes(x = 500, y = 0.0005, xend = 3000, yend = 0.0005))

dev.off()
# DONE!
```

I used this one liner to output the highest coverage contig in the velvet_repeat_lib directory:

```
grep -o -P '(?<=cov_).*(?=)' contigs.fa | sort -rn | head -n 1
```


