## Trimming and assembly

The data for Tympa and Octomys each were generated by multiplexing six RNA libraries (liver, heart, muscle, lung, kidney, and gonad (testis or ovary) on one lane.  This creates concerns about lane effects, which we will just have to include as a source of noise.

First we used Trimmomatic to trim the reads; both sets of transcriptomes were trimmed with trimmomatic version 0.32. For Tympa I ran each command line independently:

```
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Heart_log.txt Heart_R1.fastq.gz Heart_R2.fastq.gz Heart_R1_trim_paired.fastq.gz Heart_R1_trim_single.fastq.gz Heart_R2_trim_paired.fastq.gz Heart_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Kidney_log.txt Kidney_R1.fastq.gz Kidney_R2.fastq.gz Kidney_R1_trim_paired.fastq.gz Kidney_R1_trim_single.fastq.gz Kidney_R2_trim_paired.fastq.gz Kidney_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Liver_log.txt Liver_R1.fastq.gz Liver_R2.fastq.gz Liver_R1_trim_paired.fastq.gz Liver_R1_trim_single.fastq.gz Liver_R2_trim_paired.fastq.gz Liver_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Lung_log.txt Lung_R1.fastq.gz Lung_R2.fastq.gz Lung_R1_trim_paired.fastq.gz Lung_R1_trim_single.fastq.gz Lung_R2_trim_paired.fastq.gz Lung_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Muscle_log.txt Muscle_R1.fastq.gz Muscle_R2.fastq.gz Muscle_R1_trim_paired.fastq.gz Muscle_R1_trim_single.fastq.gz Muscle_R2_trim_paired.fastq.gz Muscle_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog Testis_log.txt Testis_R1.fastq.gz Testis_R2.fastq.gz Testis_R1_trim_paired.fastq.gz Testis_R1_trim_single.fastq.gz Testis_R2_trim_paired.fastq.gz Testis_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
```
 For Octomys, I wrote a bash script:

``` bash
#!/bin/bash

files=”Heart
Liver
Lung
Muscle
Kidney
Ovary”

for file in $files
do
	java -jar /work/ben/Trimmomatic-0.32/trimmomatic-0.32.jar PE -phred33 -trimlog ${file}_log.txt AO248_${file}_R1.fastq.gz AO248_${file}_R2.fastq.gz AO248_${file}_R1_trim_paired.fastq.gz AO248_${file}_R1_trim_single.fastq.gz AO248_${file}_R2_trim_paired.fastq.gz AO248_${file}_R2_trim_single.fastq.gz ILLUMINACLIP:/work/ben/Trimmomatic-0.32/adapters/trimmomatic_adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:36
done
```

I am going to use Trinity version 2.1.1 for assembly.  I am using only reads where both pairs passed the trimming step and I am going to concatenate forward and reverse reads from all libraries for each species and do the assembly jointly for all libraries for each species. This should reduce redundancy, facilitate read mapping, and hopefully simplify things in general.

On info, here is the concatenation command for the forward (R1) reads for Octomys:

```
cat Heart/AO248_Heart_R1_trim_paired.fastq Kidney/AO248_Kidney_R1_trim_paired.fastq Liver/AO248_Liver_R1_trim_paired.fastq Lung/AO248_Lung_R1_trim_paired.fastq Muscle/AO248_Muscle_R1_trim_paired.fastq Ovary/AO248_Ovary_R1_trim_paired.fastq > AO248_all_R1_trim_paired.fastq
```

and for the reverse (R2) reads:

```
cat Heart/AO248_Heart_R2_trim_paired.fastq Kidney/AO248_Kidney_R2_trim_paired.fastq Liver/AO248_Liver_R2_trim_paired.fastq Lung/AO248_Lung_R2_trim_paired.fastq Muscle/AO248_Muscle_R2_trim_paired.fastq Ovary/AO248_Ovary_R2_trim_paired.fastq > AO248_all_R2_trim_paired.fastq
```

And now the trinity assembly using version 2.1.1:

```
/home/ben/trinityrnaseq-2.1.1/Trinity --seqType fq --left AO248_all_R1_trim_paired.fastq.gz --right AO248_all_R2_trim_paired.fastq.gz --CPU 6 --max_memory 20G
```
and for the concatenated tympa transcriptomes:

```
/home/ben/trinityrnaseq-2.1.1/Trinity --seqType fq --left tympa_all_R1_trim_paired.fastq.gz --right tympa_all_R2_trim_paired.fastq.gz --CPU 6 --max_memory 20G
```

Update in July, 2016.  I am working with the Octomys and Tympanoctomys trinity assemblies again in order to provide Nate with some preliminary data for his proposal.  I have renamed the trinity.fasta files as follows: 
'/home/ben/2014_Tympanoctomys_transcriptomes/Tympano/Tympano_joint_trinity_assembly_with_concatenated_reads/trinity_out_dir/Tympa_all_transcriptomes_assembled_together.fasta'
'/home/ben/2014_Tympanoctomys_transcriptomes/Octomys/Octomys_joint_trinity_assembly_with_concatenated_reads/trinity_out_dir/Octomys_all_transcriptomes_assembled_together.fasta'

These have 280,712 and 308,854 assembled sequences for T. barrarae and O. mimax, respectively.  Then I used cd-hit to reduce redundancy as follows:
```
/home/ben/cd-hit-v4.6.1-2012-08-27/cd-hit -i Tympa_all_transcriptomes_assembled_together.fasta -o Tympa_all_transcriptomes_assembled_together_unique.fasta -c 1.00 -n 5 -M 2000
```
and 

```
/home/ben/cd-hit-v4.6.1-2012-08-27/cd-hit -i Octomys_all_transcriptomes_assembled_together.fasta -o Octomys_all_transcriptomes_assembled_together_unique.fasta -c 1.00 -n 5 -M 2000
```

While this was running, I used RepARC.pl to use a kmar approach to identify and assemble transposible elements from the raw (trimmed) sequences. This did not work initially so I tried it again with a threshold of 70 as follows:
```
./RepARK.pl -l AO248_all_R1_trim_paired.fastq.gz -l AO248_all_R2_trim_paired.fastq.gz -t 70
```
```
./RepARK.pl -l tympa_all_R1_trim_paired.fastq.gz -l tympa_all_R2_trim_paired.fastq.gz -t 70
```
This worked and made a file in a folder called 'velvet_repeat_lib' which was called 'contigs.fa' which had the repeat sequences that were constructed from the kmers.  Then this can be fed into TEclass.  I had to log into evanslab and copy the contig.fa files there because I got a perl error associated with a missing module on my account.  I was unable to use cpan to install it because of privelidges. Anyhow, here is how it was run on evanslab from within the TEclass-2.1.3 directory:
```
/home/evanslab/Hymeno_fastqc/RepArk_analysis/TEclass-2.1.3/TEclassTest.pl -r ./Octomys/contigs.fa -o ./Octomys/TE_count_and_categories.out
```
and
```
/home/evanslab/Hymeno_fastqc/RepArk_analysis/TEclass-2.1.3/TEclassTest.pl -r ./Tympa/contigs.fa -o ./Tympa/TE_count_and_categories.out
```



